### PR TITLE
fix(gevent): unload time but not typing modules for module cloning

### PR DESCRIPTION
**Pull request by Yun-Kim**
This PR makes a fix to #4863 which inadvertently removed unloading the `time` module as well as removing the `typing` module from the list of standard library modules to not unload for module cloning. We found that the `time` module is a special case (which is important to unload due to interactions with gevent patching) as it is implicitly imported by recent versions of CPython on interpreter startup, so it may be present in `sys.modules` already.
The `typing` module is also problematic on being reloaded for older versions of Python < 3.7, so we should not unload/reload this module as well.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [ ] Title is accurate.
- [ ] No unnecessary changes are introduced.
- [ ] Description motivates each change.
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Testing strategy adequately addresses listed risk(s).
- [ ] Change is maintainable (easy to change, telemetry, documentation).
- [ ] Release note makes sense to a user of the library.
- [ ] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Comment by emmettbutler** [comment]
@Yun-Kim would you mind editing the riotfile to show which additional versions of python this creates support for?

File: ddtrace/bootstrap/sitecustomize.py [hunk 1]
```diff
@@ -12,7 +12,12 @@
 import os  # noqa
 
 
-MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
+"""
+The following modules cause problems when being unloaded/reloaded in module cloning.
+Notably, unloading the atexit module will remove all registered hooks which we use for cleaning up on tracer shutdown.
+The other listed modules internally maintain some state that does not coexist well if reloaded.
+"""
+MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging", "typing"}
```
Comment (P403n1x87) [inline ddtrace/bootstrap/sitecustomize.py:20]:
```suggestion
MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
if sys.version_info < (3, 7):
    MODULES_TO_NOT_CLEANUP |= {"typing"}
```

File: riotfile.py [hunk 2]
```diff
@@ -2604,7 +2604,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
             pkgs={"requests": latest, "gevent": latest},
             venvs=[
                 Venv(
-                    pys=select_pys(min_version="3.8"),
+                    pys=select_pys(),
```
Comment (P403n1x87) [inline riotfile.py:2607]: I think we might need to keep this constraint until #5105 is merged, but let's see what the CI says.

**Review by P403n1x87 — COMMENTED** [review]
Some nits, otherwise LGTM.

File: riotfile.py (hunk 2 above)
Comment (Yun-Kim) [reply riotfile.py:2607]: Yup, just added this temporarily to verify which versions work under this change 😄

**Comment by Yun-Kim** [comment]
> @Yun-Kim would you mind editing the riotfile to show which additional versions of python this creates support for?

Looks like this change still doesn't support Python versions <= 3.7 (Hangs on 2.7, see memalloc issues on Python 3 up until 3.8+). The likely fix to support the older Python versions is contained in #5105.

File: ddtrace/bootstrap/sitecustomize.py [hunk 3]
```diff
@@ -12,7 +12,14 @@
 import os  # noqa
 
 
+"""
+The following modules cause problems when being unloaded/reloaded in module cloning.
+Notably, unloading the atexit module will remove all registered hooks which we use for cleaning up on tracer shutdown.
+The other listed modules internally maintain some state that does not coexist well if reloaded.
+"""
 MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
+if sys.version_info <= (3, 7):
```
Comment (P403n1x87) [inline ddtrace/bootstrap/sitecustomize.py:21]:
```suggestion
if sys.version_info < (3, 7):
```

Comment (P403n1x87) [reply ddtrace/bootstrap/sitecustomize.py:21]: or
```suggestion
if sys.version_info <= (3, 6):
```

**Review by emmettbutler — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**